### ISSUE TITLE
`sam build` handles when some functions/layers already built by other methods (with CodeUri pointing to local build artifact ZIPs)

### ISSUE THREAD

**Issue by mfriesen** [aws/aws-sam-cli#2703]
### Describe your idea/feature/enhancement

Previous to sam cli 1.20 you could specify a Lambda zip file in AWS::Serverless::Function CodeUri and sam would build and create an S3 link for the lambda zip file. (https://github.com/aws/aws-sam-cli/issues/2693)

As of 1.20, sam only supports local directories where it try and build the software or S3 urls. This really limits the use cases for sam for building large enterprise systems or being able to reuse code without copy and pasting it into every project that uses the same "module".

### Proposal

The AWS::Serverless::Function CodeUri parameter should support local Zip files. When building it should upload the zip file to the SAM cli S3 bucket and replace the CodeUri with the S3 link.

### Edit [by @aahung] 
The issue description does not reflect the root issue. For clarification, please see the 1st comment below.

**Comment by aahung** [comment]
AWS::Serverless::Function always support CodeUri parameter as a ZIP file, before or after 1.20.0. 

@mfriesen let me put some clarification here, following the discussion from #2693:

**First, there are two type of ZIP files, source code used by SAM to build into artifact and build artifacts** Here the ZIP file referred by @mfriesen is the latter. 

Some background of @mfriesen's template: it is a template with nested stacks in it:

- root template does not have buildable resources (i.e., local AWS::Serverless::Function or AWS::Serverless::LayerVersion).
- nested stacks contains functions, but @mfriesen already build them using other approach than `sam build`, that's why a zip file is used.
- the current workflow is
    1. `sam build`
    2. `sam package` or `sam deploy`

Before 1.20.0, the step 1 (`sam build`) actually does not build anything excepting moving all the root template file to `.aws-sam/build`. now in 1.20.0, with nested stack support, `sam build` tries to build functions in nested stacks, that are already built, which caused error. 

_In short, 1.20.0 does not change the behavior on ZIP file handling, it just tries to build resources which was left behind before. `sam build` always supports CodeUri to be a zip file, just not local artifact-type zip file._

Since the root template does not have buildable resources, the original suggestion given to @mfriesen was to skip step 1, and the issue was resolved.

As a follow up discussin in ##2693, @mfriesen came up with a new scenario, what if in the template, some functions use artifact-type ZIP as `CodeUri` and some functions use source code `CodeUri`? How SAM CLI should handle this scenario?

**Comment by aahung** [comment]
@mfriesen I am gonna change the title of this issue to make it clearer, if I misinterpret your issue, please let me know.

**Comment by aahung** [comment]
### Options for SAM users

- upload the build artifact ZIPs to S3 and use `s3://...` as `CodeUri`, in that case, SAM CLI will skip them.

### Possible Options for SAM CLI Future Versions

- Automatically detect whether a ZIP is source code to build and artifacts
- Allow customers to specify which functions/resources to skip by either introduce a metadata property or CLI

**Comment by pauldes** [comment]
>   upload the build artifact ZIPs to S3 and use `s3://...` as `CodeUri`, in that case, SAM CLI will skip them.

Actually, it looks like SAM CLI (v1.21.1) will not skip them, no matter if :
- `CodeUri` is a S3 object URI or a path to a local .zip file
- `Type` is `AWS::Serverless::Function` or `AWS::Lambda::Function`

As a consequence I cannot figure out a solution to use `sam build` with an existing zip file. Will https://github.com/aws/aws-sam-cli/pull/2737 solve this ?

**Comment by aahung** [comment]
> > upload the build artifact ZIPs to S3 and use `s3://...` as `CodeUri`, in that case, SAM CLI will skip them.
> 
> Actually, it looks like SAM CLI (v1.21.1) will not skip them, no matter if :
> 
>     * `CodeUri` is a S3 object URI or a path to a local .zip file
> 
>     * `Type` is `AWS::Serverless::Function` or `AWS::Lambda::Function`
> 
> 
> As a consequence I cannot figure out a way to use `sam build` with an existing zip file. Will #2737 solve this ?

Yes, sorry my original suggestion was not accurate. We are working on implementing skipping s3 and zip files.

**Comment by aahung** [comment]
This has been implemented [1.22.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.22.0)
